c10/core/SymInt.cpp

```diff
@@ -29,23 +51,20 @@ SymInt SymInt::operator*(SymInt sci) const {
   if (!is_symbolic() && !sci.is_symbolic()) {
     return SymInt(data_ * sci.data_);
   }
-  // TODO: This is way to much boilerplate
-  std::shared_ptr<SymbolicIntNode> a =
-      is_symbolic() ? toSymbolicIntNode() : nullptr;
-  std::shared_ptr<SymbolicIntNode> b =
-      sci.is_symbolic() ? sci.toSymbolicIntNode() : nullptr;
+  auto res = normalize_symints(*this, sci);
+  return SymInt::toSymInt(res[0]->mul(res[1]));
+}
 
-  SymbolicIntNode* common = a ? a.get() : b.get();
-  // TODO: technically we need to check that the classes match
-  if (!a) {
-    a = common->wrap(data_);
-    toSymInt(a); //
-  }
-  if (!b) {
-    b = common->wrap(sci.data_);
-    toSymInt(b);
+bool SymInt::operator==(SymInt sci) const {
+  if (!is_symbolic() && !sci.is_symbolic()) {
+    return data_ == sci.data_;
   }
-  return SymInt::toSymInt(a->add(b));
+  auto res = normalize_symints(*this, sci);
+  return res[0]->eq(res[1])->bool_();
+}
+
+bool SymInt::operator!=(SymInt sci) const {
+  return !(*this == sci);
```

Comment:
I told @Krovatkin that we needed to define neq directly, but now I've forgotten why

Reply:
Perhaps since torchinductor does some special simplifications for `==` vs. `!=`?